src/project-converter.ts

```diff
@@ -0,0 +1,165 @@
+    }
+  }
+
+  /**
+   * Check if a document is explicitly excluded or has already been converted
+   * to decide if it should be converted or skipped.
+   */
+  shouldConvertDocument(document: Document): boolean {
+    const documentUrl = getDocumentUrl(document);
+    if (this.conversionResults.has(documentUrl)) {
+      return false;
+    }
+    if (this.conversionSettings.excludes.has(documentUrl)) {
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Specifically convert an HTML document to a JS module. Useful during
+   * conversion for dependencies where the type of result is explictly expected.
+   */
+  convertDocumentToJs(document: Document, visited: Set<OriginalDocumentUrl>) {
+    if (!this.shouldConvertDocument(document)) {
+      return;
+    }
+    const documentConverter = new DocumentConverter(this, document, visited);
+    const newModule = documentConverter.convertToJsModule();
+    this._handleConversionResult(newModule);
+  }
+
+  /**
+   * Specifically convert an HTML document without changing the file type
+   * (changes imports and inline scripts to modules as necessary.) Useful during
+   * conversion for dependencies where the type of result is explictly expected.
+   */
+  convertDocumentToHtml(document: Document, visited: Set<OriginalDocumentUrl>) {
+    if (!this.shouldConvertDocument(document)) {
+      return;
+    }
+    const documentConverter = new DocumentConverter(this, document, visited);
+    const newModule = documentConverter.convertAsToplevelHtmlDocument();
+    this._handleConversionResult(newModule);
+  }
+
+  /**
+   * A private instance method for handling new conversion results, exports,
+   * etc.
+   */
+  private _handleConversionResult(newModule: ConversionResult): void {
+    this.conversionResults.set(newModule.originalUrl, newModule);
+    if (newModule.output.type === 'js-module') {
+      for (const expr of newModule.output.exportedNamespaceMembers) {
+        this.namespacedExports.set(
+            expr.oldNamespacedName,
+            new JsExport(newModule.convertedUrl, expr.es6ExportName));
+      }
+    }
+  }
+
+  /**
+   * This method collects the results after all documents are converted. It
+   * handles out some broken edge-cases (ex: shadycss) and sets empty map
+   * entries for files to be deleted.
+   */
+  getResults(): Map<ConvertedDocumentFilePath, string|undefined> {
+    const results = new Map<ConvertedDocumentFilePath, string|undefined>();
+
+    for (const convertedModule of this.conversionResults.values()) {
+      // These files were already ES6 modules, so don't write our conversions.
+      // Can't be handled in `excludes` because we want to preserve imports of.
+      // TODO(fks): This is hacky, ProjectConverter isn't supposed to know about
+      //  project layout / file location. Move into URLHandler, potentially make
+      //  its own `excludes`-like settings option.
+      if (convertedModule.originalUrl ===
```

Comment:
With 4 items, I'd go ahead and make a Set containing these, even if it's going to change in the future. Also, can we just detect that these are modules and not convert them?

Reply:
Conversion and analysis of a module (what are its exports, namespaces, etc) are coupled at this point. Can't do one without the other.